custom/icds_reports/migrations/0104_citus_composite_key.py

```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+from django.apps import apps
+from django.db import connections, migrations
+from django.conf import settings
+
+from corehq.sql_db.connections import get_icds_ucr_db_alias
+
+
+def _citus_composite_key_sql(model_cls):
+    connection = connections[get_icds_ucr_db_alias()]
+    with connection.schema_editor() as schema_editor:
+        pkey_name = '{}_pkey'.format(model_cls._meta.db_table)
+        fields = list(model_cls._meta.unique_together)[0]
+        columns = [model_cls._meta.get_field(field).column for field in fields]
+        index = schema_editor._create_index_name(model_cls, columns, suffix="_uniq")
+        sql = """
+            CREATE UNIQUE INDEX {index} on "{table}" ({cols});
+            ALTER TABLE "{table}" DROP CONSTRAINT {pkey_name},
+            ADD CONSTRAINT {pkey_name} PRIMARY KEY USING INDEX {index};
+        """.format(
+            table=model_cls._meta.db_table,
+            pkey_name=pkey_name,
+            index=index,
+            cols=','.join(columns)
+        )
+        reverse_sql = """
+            ALTER TABLE "{table}" DROP CONSTRAINT IF EXISTS {index},
+            DROP CONSTRAINT IF EXISTS {pkey_name},
+            ADD CONSTRAINT {pkey_name} PRIMARY KEY ({pkey});
+            DROP INDEX IF EXISTS {index};
+        """.format(
+            table=model_cls._meta.db_table,
+            pkey_name=pkey_name,
+            pkey=model_cls._meta.pk.name,
+            index=index,
+        )
+    if getattr(settings, 'UNIT_TESTING', False):
+        return sql, reverse_sql
+    else:
+        return migrations.RunSQL.noop, reverse_sql
+
+
+def get_sql_operations():
+    models_to_update = [
+        'aggregatebirthpreparednesforms',
+        'aggregateccsrecorddeliveryforms',
+        'aggregateccsrecordpostnatalcareforms',
```

Comment:
Has this been checked against currently existing tables with partitions? I think that constraints aren't automatically propagated to child tables. I'm not 100% sure if that differs for PRIMARY KEY, but I don't believe that it is any different.

Reply:
In fact, it looks like it doesn't get propagated to child tables, because of postgres's limitation of inheritance on unique/primary-key constraints and index. See caveats under https://www.postgresql.org/docs/9.5/ddl-inherit.html. This means we will probably have to run similar SQL on all child tables? Is there any other better way?

Reply:
(I had tested this manually on a local pg db)

Reply:
Since this is not going to be run on the live site (only for test setup), I do not believe the propagation to the child tables is an issue. 

For any tables that remain partitioned in future converting them to use PG native partitioning will result in the primary key being propagated however it also requires that the partition column be in the primary key (I don't think I realised that before and not sure how I got around it before but seems like extra reason not to have tables partitioned).